firebase-database/src/androidTest/java/com/google/firebase/database/QueryTest.java

```diff
@@ -3401,6 +3407,140 @@ public void onComplete(DatabaseError error, DatabaseReference ref) {
+            .build(),
+        name);
+  }
+
+  @Test
+  public void emptyQueryGet() throws DatabaseException, InterruptedException {
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getNamespace(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    try {
+      Tasks.await(db.getReference("dummy/").setValue(42L));
+      assertNull(Tasks.await(db.getReference("null/").get()).getValue());
+    } catch (ExecutionException e) {
+      fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void offlineQueryGet() throws DatabaseException, InterruptedException {
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getNamespace(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    DatabaseReference node = db.getReference();
+    try {
+      Tasks.await(node.setValue(42L));
+    } catch (ExecutionException e) {
+      fail();
+    }
+    db.goOffline();
+    try {
+      Tasks.await(node.get());
+    } catch (ExecutionException e) {
+      assertEquals(e.getCause().getMessage(), "Client is offline");
+      return;
+    }
+    fail("Client get succeeded even though offline.");
+  }
+
+  @Test
+  public void getQueryBasic() throws DatabaseException, InterruptedException {
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getNamespace(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    DatabaseReference node = db.getReference();
+    try {
+      Tasks.await(node.setValue(42));
+      assertEquals(42L, Tasks.await(node.get()).getValue());
+    } catch (ExecutionException e) {
+      fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void getQueryCached()
+      throws DatabaseException, InterruptedException, TimeoutException, TestFailure {
+    FirebaseApp app =
+        appForDatabaseUrl(IntegrationTestValues.getAltNamespace(), UUID.randomUUID().toString());
+    FirebaseDatabase db = FirebaseDatabase.getInstance(app);
+    DatabaseReference ref = db.getReference();
+    final Semaphore semaphore = new Semaphore(0);
+    ValueEventListener listener =
+        new ValueEventListener() {
+          @Override
+          public void onDataChange(@NonNull DataSnapshot snapshot) {
+            if (snapshot.getValue() != null && snapshot.getValue().equals(42L)) {
+              semaphore.release();
+            }
+          }
+
+          @Override
+          public void onCancelled(@NonNull DatabaseError error) {}
+        };
+    ref.addValueEventListener(listener);
+    ref.setValue(42L);
+    IntegrationTestHelpers.waitFor(semaphore);
+    db.goOffline();
+    try {
+      // Since we still have a listener on `ref`, the 42L should be cached here.
+      assertEquals(42L, Tasks.await(ref.get()).getValue());
+    } catch (ExecutionException e) {
+      fail(e.getMessage());
+    } finally {
+      ref.removeEventListener(listener);
+    }
```

Comment:
Please also add a test that shows the the value is still cached even when there are no remaining listeners. The SDK should keep up to 10 MB of least recently used data. For this to work, you have to track the query via the persistence manager. See here: https://github.com/firebase/firebase-android-sdk/blob/d92096257b3e750c72757fc7c4bc9d7393db2046/firebase-database/src/main/java/com/google/firebase/database/core/SyncTree.java#L505